### PR TITLE
Add two apps: Bernoulli gates and Euclidean rhythm

### PR DESCRIPTION
**Pull request by Bridgee**
Two new apps:
1. Bernoulli gates
A dual Bernoulli gates motivated by Mutable Instruments Branches.
2. Euclidean rhythm
Six independent Euclidean gate generators with internal/external clock and cv controllability.

### CODE REVIEW

**Comment by Bridgee** [comment]
They are only working as the stand-alone app now, and I'm working on wrapping them at the EuroPiScript class now. But you can always test the apps by changing the name to main.py

**Comment by Bridgee** [comment]
They are now inheriting the EuroPiScript class and work well with the new menu function.

File: software/contrib/euclidean_rhythm.py
Comment (mjaskula) [inline software/contrib/euclidean_rhythm.py:91]: If you give your flag variables slightly more descriptive names you won't need such verbose comments, and the code will be more readable where they are used.

For example, if this var was instead something like `self.external_clock = False` then the usage on line 224 could be `if self.external_clock:` and the reader wouldn't need to remember which value represented which mode.

File: software/contrib/euclidean_rhythm.py
Comment (mjaskula) [inline software/contrib/euclidean_rhythm.py:101]: I think that this entire `if`/`elif` block could be replaced with:

`self.notes_para[self.selection_idx][self.editing_idx] -= 1`

The same change follows in the 'add' section below.

File: software/contrib/euclidean_rhythm.py
Comment (mjaskula) [inline software/contrib/euclidean_rhythm.py:159]: It's good practice to add a message to an error in order to make it easier to debug when it occurs. For example:

raise ValueError(f"pulses ({pulses}) unexpectedly larger than steps ({steps})")

File: software/contrib/euclidean_rhythm.py
Comment (mjaskula) [inline software/contrib/euclidean_rhythm.py:208]: The if structure above can be simplified such that the line to draw each rect only needs to occur once. The first box always gets drawn. Each subsequent box gets drawn for every larger index. So I think this would work:

```python
oled.fill_rect(box_position[selection_idx][0]+29, box_position[selection_idx][1]+1, 2, 2, 1)
if editing_idx >= 1:
   oled.fill_rect(box_position[selection_idx][0]+32, box_position[selection_idx][1]+1, 2, 2, 1)
if editing_idx >= 2:
   oled.fill_rect(box_position[selection_idx][0]+35, box_position[selection_idx][1]+1, 2, 2, 1)
if editing_idx >= 3:
   oled.fill_rect(box_position[selection_idx][0]+38, box_position[selection_idx][1]+1, 2, 2, 1)
```

Besides smaller code, this has the advantage that in the event you want to change these rects you only need to change it in one place.

**Review by mjaskula — COMMENTED** [review]
These scripts look really great, I can't wait to try them out. I've looked over the code a bit and left some suggestions on some minor code improvements. I don't feel like I quite understand the Euclidean script quite yet, so I'm going to look over that some more, but I'm out of time for today.

Please don't take any of these as comments as criticism. I think of PRs as conversations and learning experiences for both the author and the reviewer. With the goal of arriving at the best possible code being added to the repository.

**Comment by Bridgee** [comment]
> These scripts look really great, I can't wait to try them out. I've looked over the code a bit and left some suggestions on some minor code improvements. I don't feel like I quite understand the Euclidean script quite yet, so I'm going to look over that some more, but I'm out of time for today.
> 
> Please don't take any of these as comments as criticism. I think of PRs as conversations and learning experiences for both the author and the reviewer. With the goal of arriving at the best possible code being added to the repository.

Hey mjaskula,
Thanks for all those valuable comments. I'll go through them as soon I get some time.
For the Euclidean script, there's known issue:
The control and screen refresh are syncronized with the tempo. By default, it uses the internal clock, and there's no problem. However, if you switch to external clock, and there's no incoming trigger from din, the system will "freeze" until the pulse received.

**Comment by awonak** [comment]
> However, if you switch to external clock, and there's no incoming trigger from din, the system will "freeze" until the pulse received.

I also bumped into this issue a while back. I experimented a bit with a solution linked below that might offer some inspiration.

When you're in external clock mode, the script hangs and waits for a clock pulse. Instead of hanging indefinitely, you could put a "timeout" on that external wait behavior. If 3000ms has elapsed (the distance between quarter notes at 20 BPM), then automatically switch over to internal clock mode. 

You might want to make this behavior configurable, because there may be some scenarios where you want to use Euclidean Rhythms with trigger sources other than a clock :)

https://github.com/awonak/EuroPi/blob/awonak-coin-toss-ext-clock/contrib/clock.py#L98-L103

File: software/contrib/euclidean_rhythm.py
Comment (awonak) [reply software/contrib/euclidean_rhythm.py:159]: I would suggest avoiding throwing an exception, which terminates the script, and instead use the [clamp](https://allen-synthesis.github.io/EuroPi/generated/europi.html#europi.clamp) function. Something like this:

```python
pulses = clamp(int(pulses), 0, steps)
```

File: software/contrib/bernoulli_gates.py
Comment (Bridgee) [reply software/contrib/bernoulli_gates.py:23]: I put it to __main__, so that it only work when it runs as a stand-alone.

File: software/contrib/bernoulli_gates.py
Comment (Bridgee) [reply software/contrib/bernoulli_gates.py:28]: Added a SingleBernoulliGate class so that the code can be reused from the two gates.

File: software/contrib/euclidean_rhythm.py
Comment (Bridgee) [reply software/contrib/euclidean_rhythm.py:28]: Changed

File: software/contrib/euclidean_rhythm.md
Comment (Bridgee) [reply software/contrib/euclidean_rhythm.md]: Changed

**Comment by Bridgee** [comment]
> > However, if you switch to external clock, and there's no incoming trigger from din, the system will "freeze" until the pulse received.
> 
> I also bumped into this issue a while back. I experimented a bit with a solution linked below that might offer some inspiration.
> 
> When you're in external clock mode, the script hangs and waits for a clock pulse. Instead of hanging indefinitely, you could put a "timeout" on that external wait behavior. If 3000ms has elapsed (the distance between quarter notes at 20 BPM), then automatically switch over to internal clock mode.
> 
> You might want to make this behavior configurable, because there may be some scenarios where you want to use Euclidean Rhythms with trigger sources other than a clock :)
> 
> https://github.com/awonak/EuroPi/blob/awonak-coin-toss-ext-clock/contrib/clock.py#L98-L103

Sounds like a good idea. I'll try to apply this later

File: software/contrib/euclidean_rhythm.py
Comment (Bridgee) [reply software/contrib/euclidean_rhythm.py:155]: This code block is directly copied from the python implementation of the Bjorklund algorithm repository. I simply delete this part as the input of this function is already bounded by other logic.

File: software/contrib/euclidean_rhythm.py
Comment (Bridgee) [reply software/contrib/euclidean_rhythm.py:97]: Changed

File: software/contrib/euclidean_rhythm.py
Comment (Bridgee) [reply software/contrib/euclidean_rhythm.py]: Changed.

File: software/contrib/bernoulli_gates.py
Comment (Bridgee) [reply software/contrib/bernoulli_gates.py]: They are all in the SingleBernoulliGate now.

File: software/contrib/euclidean_rhythm.py
Comment (Bridgee) [reply software/contrib/euclidean_rhythm.py:87]: Changed

File: software/contrib/bernoulli_gates.py
Comment (mjaskula) [reply software/contrib/bernoulli_gates.py]: Sorry, this isn't a menu default, it's a eurpoi library default. See lines [51](https://github.com/Allen-Synthesis/EuroPi/blob/main/software/firmware/europi.py#L51) and 86](https://github.com/Allen-Synthesis/EuroPi/blob/main/software/firmware/europi.py#L86). So it doesn't matter where you put the call to `set_samples()`, it's not really doing much. In fact, I think it's even more confusing to only do it sometimes (that is only when run stand alone).

File: software/contrib/bernoulli_gates.py
Comment (mjaskula) [reply software/contrib/bernoulli_gates.py:124]: I think this is an improvement. Nice work.

File: software/contrib/bernoulli_gates.py
Comment (mjaskula) [inline software/contrib/bernoulli_gates.py:27]: This `id` parameter is fine for now, but if we ever wanted to add a third gate I think we would want to make all of these fields constructor params. (For example if an expander module added the right ins and outs to support a third gate.) The idea here would be to separate the responsibilities of 'how a gate works' from 'which i/o it uses and where on the screen it displays'. 

If we get the bernoulli gate class isolated enough it could be possible to pull it up into a library class where other scripts could build upon it. I'm not saying that you should be thinking about future uses like this when writing scripts, I'm just throwing ideas out that people might be interested in running with.

File: software/contrib/bernoulli_gates.py
Comment (awonak) [inline software/contrib/bernoulli_gates.py:105]: Minor, but important to fix before we merge; the README refers to this as labeled "L", but the oled display shows "G" when in gate/latch mode. Either is label is fine, but make sure they're consistent.

File: software/contrib/bernoulli_gates.py
Comment (awonak) [inline software/contrib/bernoulli_gates.py:100]: One more important point to consider; when each bernoulli gate sleeps independently, that will cause a noticeable "swing" like delay between each trigger. To fix this, instead of calling sleep here, return a boolean indicating whether or not a trigger should occur. Then after both `regular_maneuver` functions have been called, call sleep there based on the results of both function calls.

```python
    def regular_maneuver(self):
        triggered = false
        if self.mode_flg == 0:
            oled.text('Tr', self.text1_pos, OLED_HEIGHT-8, 1)
            triggered = true
        ....
        return triggered
    
```
```python
    # Regular maneuver
    triggered1 = self.first_gate.regular_maneuver()
    triggered2 = self.second_gate.regular_maneuver()
    if triggered1 or triggered2:
        sleep_ms(TRG_T)
```

**Review by awonak — COMMENTED** [review]
I just played around with the Bernoulli Gates script and I think the UI is fantastic! Well done, and thank you for  building this script!

**Comment by awonak** [comment]
@Bridgee It looks like we're pretty close to having the Bernoulli Gates script ready for merge. I think the Euclidean Rhythm script is going to need some more time for review. Would you mind splitting that script off into a separate branch? Or if you don't mind the delay, we can keep moving forward with this pull request.

File: software/contrib/bernoulli_gates.py
Comment (Bridgee) [reply software/contrib/bernoulli_gates.py]: Removed

File: software/contrib/bernoulli_gates.py
Comment (Bridgee) [reply software/contrib/bernoulli_gates.py:16]: Moved

File: software/contrib/bernoulli_gates.py
Comment (Bridgee) [reply software/contrib/bernoulli_gates.py:105]: Changed the README to keep the consistency.

File: software/contrib/bernoulli_gates.py
Comment (Bridgee) [reply software/contrib/bernoulli_gates.py]: Thanks for the advice, please see my modification.

File: software/contrib/bernoulli_gates.py
Comment (Bridgee) [reply software/contrib/bernoulli_gates.py]: Good point! I move the sleep_ms function to the main loop instead of keeping them in the regular_maneuver() function.

**Comment by Bridgee** [comment]
> @Bridgee It looks like we're pretty close to having the Bernoulli Gates script ready for merge. I think the Euclidean Rhythm script is going to need some more time for review. Would you mind splitting that script off into a separate branch? Or if you don't mind the delay, we can keep moving forward with this pull request.

I just refined the Bernoulli Gates script based on all of your comments. As to the Euclidean Rhythm script, please feel free to play with it and let me know your thought. I don't mind waiting until both of them are ready to be merged.

File: software/contrib/bernoulli_gates.py
Comment (mjaskula) [inline software/contrib/bernoulli_gates.py:49]: Typically methods whose names start with `get` return values. A name like `update_prob` might be more clear.

File: software/contrib/bernoulli_gates.py
Comment (mjaskula) [inline software/contrib/bernoulli_gates.py:24]: I like the way you made the `port3` stuff flexible here. It's much more elegant than the method that I was thinking of suggesting.

Could you add `port3_source_cv` to the docstring here?

**Review by mjaskula — COMMENTED** [review]
I added a few more comments on the bernoulli_gates script with some potential readability and complexity changes. These are minor, (take those that make sense to you and ignore the rest) and even without them I think that the script is ready to merge.

I did notice one undocumented parameter to `SingleBernoulliGate.__init__()` that should be added.

Overall nice work and a great addition to the suite of scripts. I'll tackle the Euclidean script in the coming days.

**Review by awonak — COMMENTED** [review]
Nice work with the Bernoulli Gates script. I'm starting to dig into the Euclidean Rhythm script and is see a lot of room for improvement. You did a great job with the feedback from the Bernoulli Gates script, so I'm confident we can get things to an improved state here too.

Overall, there is a lot of repetition and duplication of work. If we introduce some reuse patterns, that will make the code cleaner, easier to read and debug, and run smoother. Additionally, I think having all 6 rhythm patterns displayed on the OLED at once is tricky since there is such little space to work with. I'd suggest thinking of ways to improve the usability and clarity (perhaps only show one pattern at a time?).

First, I'd like you to take a look at the Euclidean Rhythms script I wrote for the EuroPi Prototype. The concept here is relatively simple. The `EuclideanRhythm` class holds 4 instances of `Pattern`s. The main script class handles interactions with buttons and knobs, and passes those to the current selected `Pattern` to update its state. The `Pattern` class only knows about its own state in terms of steps, pulses, and the current step in the pattern. 

https://github.com/awonak/EuroPiAlt/blob/main/src/scripts/euclidean_rhythm.py

I'd like you to think about the instances in your code where there is a lot of repetition or if/else conditions and how you can reduce that repetition by introducing a class to abstract away behavior, or add a helper method to reuse code.

If you'd like a little more hands on help cleaning this up, feel free to reach out to Matt (mjaskula#7507) or me (Slumbernemy#0357) on Discord and we can dig in together to improve this script.

**Comment by Bridgee** [comment]
> Nice work with the Bernoulli Gates script. I'm starting to dig into the Euclidean Rhythm script and is see a lot of room for improvement. You did a great job with the feedback from the Bernoulli Gates script, so I'm confident we can get things to an improved state here too.
> 
> Overall, there is a lot of repetition and duplication of work. If we introduce some reuse patterns, that will make the code cleaner, easier to read and debug, and run smoother. Additionally, I think having all 6 rhythm patterns displayed on the OLED at once is tricky since there is such little space to work with. I'd suggest thinking of ways to improve the usability and clarity (perhaps only show one pattern at a time?).
> 
> First, I'd like you to take a look at the Euclidean Rhythms script I wrote for the EuroPi Prototype. The concept here is relatively simple. The `EuclideanRhythm` class holds 4 instances of `Pattern`s. The main script class handles interactions with buttons and knobs, and passes those to the current selected `Pattern` to update its state. The `Pattern` class only knows about its own state in terms of steps, pulses, and the current step in the pattern.
> 
> https://github.com/awonak/EuroPiAlt/blob/main/src/scripts/euclidean_rhythm.py
> 
> I'd like you to think about the instances in your code where there is a lot of repetition or if/else conditions and how you can reduce that repetition by introducing a class to abstract away behavior, or add a helper method to reuse code.
> 
> If you'd like a little more hands on help cleaning this up, feel free to reach out to Matt (mjaskula#7507) or me (Slumbernemy#0357) on Discord and we can dig in together to improve this script.

Hi @awonak,
Thanks for your comments. I've cleaned up my code as suggested. I've tried to make everything instances or functions to avoid redundancy.
BTW, I'm also working on adding a 'reset steps' function. 
Bridge

File: software/contrib/bernoulli_gates.py
Comment (Bridgee) [reply software/contrib/bernoulli_gates.py]: I kept the default parameters in the definition part, but still gave both of the instances the actual input. And, yes, it's easier to read and understand the configuration.

File: software/contrib/euclidean_rhythm.py
Comment (Bridgee) [reply software/contrib/euclidean_rhythm.py]: Improved.

**Comment by awonak** [comment]
Hey @Bridgee, I wanted to check in and let you know that I'm still working on feedback for this script. I have some suggestions on how to further improve things and it's taking me longer than I'd like. Great work so far, and thanks for your patience as we get these scripts up to par!

**Comment by awonak** [comment]
@Bridgee would you be willing to remove the Euclidean Rhythm scrips out of this PR? I think the Benroulli Gates script is good to go and I'd love to get that scrip added to the contrib scripts and revisit Euclidean Rhythm separately.

**Comment by Bridgee** [comment]
> @Bridgee would you be willing to remove the Euclidean Rhythm scrips out of this PR? I think the Benroulli Gates script is good to go and I'd love to get that scrip added to the contrib scripts and revisit Euclidean Rhythm separately.

Hi @awonak, sorry for the late reply. I was trapped in my own work recently. I removed the Euclidean Rhythm from the current PR just now. Please go ahead and merge it. I'll create another PR for the ER module later.

File: software/contrib/menu.py
Comment (awonak) [inline software/contrib/menu.py:5]: One last request, please remove the Euclidean Rhythm references from the menu and we're good to go!

Comment (Bridgee) [reply software/contrib/menu.py]: Removed.

**Review by mjaskula — APPROVED** [review]

File: software/contrib/euclidean_rhythm.md
Comment (Bridgee) [reply software/contrib/euclidean_rhythm.md]: Right, it's a typo

Comment (awonak) [reply software/contrib/euclidean_rhythm.md]: I'm planning on removing the Coin Toss script in favor of this script. I'll add this fix when I submit that PR.

**Pull request merged**
